apiserver/facades/client/cloud/cloud_test.go

```diff
@@ -941,16 +1218,39 @@ func (s *cloudSuite) TestUpdateCredentialsSomeModelsFailedValidationForce(c *gc.
 }
 
 func (s *cloudSuite) TestUpdateCredentialsAllModelsFailedValidation(c *gc.C) {
-	s.backend.credentialModelsF = func(tag names.CloudCredentialTag) (map[string]string, error) {
-		return map[string]string{
-			coretesting.ModelTag.Id():              "testModel1",
-			"deadbeef-2f18-4fd2-967d-db9663db7bea": "testModel2",
-		}, nil
+	adminTag := names.NewUserTag("admin")
+	defer s.setup(c, adminTag).Finish()
+
+	model := map[string]string{
+		coretesting.ModelTag.Id():              "testModel1",
+		"deadbeef-2f18-4fd2-967d-db9663db7bea": "testModel2",
+	}
+	aCloud := jujucloud.Cloud{
+		Name:      "dummy",
+		Type:      "dummy",
+		AuthTypes: []jujucloud.AuthType{jujucloud.EmptyAuthType, jujucloud.UserPassAuthType},
+		Regions:   []jujucloud.Region{{Name: "nether", Endpoint: "endpoint"}},
 	}
 
-	s.PatchValue(cloudfacade.ValidateNewCredentialForModelFunc, func(backend credentialcommon.PersistentBackend, callCtx context.ProviderCallContext, credentialTag names.CloudCredentialTag, credential *cloud.Credential, migrating bool) (params.ErrorResults, error) {
-		return params.ErrorResults{[]params.ErrorResult{{&params.Error{Message: "not valid for model"}}}}, nil
-	})
+	modelBackend := newModelBackend(c, aCloud,
+		coretesting.ModelTag.Id())
+	tag, err := cloudCredentialTag("three", "julia", "meep", map[string]string{}).CloudCredentialTag()
+	c.Assert(err, jc.ErrorIsNil)
+
+	s.PatchValue(cloudfacade.ValidateNewCredentialForModelFunc,
+		func(
+			backend credentialcommon.PersistentBackend, callCtx context.ProviderCallContext,
+			credentialTag names.CloudCredentialTag, credential *jujucloud.Credential, migrating bool,
+		) (params.ErrorResults, error) {
+			return params.ErrorResults{[]params.ErrorResult{{&params.Error{Message: "not valid for model"}}}}, nil
+		})
+
+	backend := s.backend.EXPECT()
+	backend.CredentialModels(tag).Return(model, nil)
+
+	pool := s.pool.EXPECT()
+	pool.GetModelCallContext(gomock.Any()).Return(modelBackend, context.NewEmptyCloudCallContext(),
+		errors.New("not valid for model")).Times(2)
```

Comment:
I Agree. I don't think we need the DoAndReturn. It will force extra lines passing the function argument instead I chose to just use `Return` and pass in there the expected return types in one line.